### DISCUSSION TITLE
Is the content safe contains within a long URL, without any password to open this URL?

### DISCUSSION THREAD

**Discussion by rockman413** [PrivateBin/PrivateBin#1270] [Q&A]
### Describe the problem/question

Hi there,
   Can someone try all kinds of combination of letters as the URL to open every possible PrivateBin pastes?
    My friend send me a long URL of PrivateBin privately, when I click the link it doesn't require a password to open it.
   Is the content within safe? Technically only my friend and I have it but I assume anyone can "crawl" out a workable URL and get the content inside a working URL?

### Did you use the FAQ section?

- [X] Yes, I have read [the FAQ](https://github.com/PrivateBin/PrivateBin/wiki/FAQ) and I found no solution/answer there.

### What you did?

1.
2.
...

### Issue reproducibility

No, I cannot reproduce it on https://privatebin.net.

**Comment by elrido** [comment]
See also: https://sebsauvage.net/wiki/doku.php?id=php:zerobin#how_does_it_work

In short, that long URL contains two pieces:

1. identifier - 8 bytes long
2. key - 32 bytes long

The identifier is a Fowler–Noll–Vo 1a hash of the submitted paste (used to be md5, which was controversial/misleading). It only serves to avoid that the same content is uploaded twice (at the same time). If a paste with the same hash already exists, it is rejected. It is only 8 bytes long, so with "only" 2^64 requests, all possible pastes of an instance can be retrieved. This is not impossible, but would likely take many years if not parallelized massively. Such an operation would likely trip up any DDoS protections of the providers that host your instance and get blocked on that end.

But the pastes stored / retrieved contain an encrypted payload. To actually decrypt & read the contents of that message, the key is required. It is a cryptographic random number, generated by the paste creator. Assuming no additional password is set and no weakness in AES being discovered, it could be "brute-forced", that is discovered by trial and error, by trying to decrypt the message at most 2^256 times. At this time we would assume that to take a very long time, longer than either of us will likely remain alive.

So to your question, "is the content safe": Your content should be hidden, as long as the URL is not shared/intercepted with/by unwanted parties. *But* the meta-data of the paste is not hidden - that includes the time of the paste creation, which IP created it, which IP accessed it and when, etc.

You can increase confidentiality by: Creating pastes as "burn-after-reading", so it can only be opened once and immediately gets deleted (less risk for third parties finding it by walking the identifiers), creating it with a password set or using a different tool / form of encrypted communication.

**Reply by rockman413** [reply]
Thanks a lot!
1.  So in short, am I correct that any kind of "crawl" such as to try different combination of letters in URL to visit, will likely trip a block as DDoS protections?  Further more, If someone shares a file via a iCloud Link, that file is also safe as multiple try different URL to visit will trip the DDoS protection too?
2.  "the meta-data of the paste is not hidden - that includes the time of the paste creation, which IP created it, which IP accessed it and when, etc."--> so how can someone else know the time of the creation? which IP created it? why it's not hidden? how can someone even know the existence of this paste?

**Reply by elrido** [reply]
1. It depends on the nature of the crawl and the host in question. I tried to point out that such an exhaustive crawl would take years if done naively, single-threaded from a single host. When scaling this process to reduce this time, you then likely trip up DDoS protections or simply exhaust the hosts uplink. This is for the likelihood of someone guessing your exact shared link. I do not know what scheme iCloud uses, but such hard to guess identifiers are a very common practice - it is what gets used in NextCloud or Office 365 when you generate an "anyone with the link can read this file"-link. Unlike PrivateBin, those services will probably use a cryptographic hash function and longer identifiers, since that ID will be their only access protection.
2. Network traffic monitoring. Unfortunately we live in a world where that is no longer a hypothetical scenario, but actual common practice, even in most "western democracies". We have to assume anyone snooping on the traffic patterns on the instance in question can track the IPs of sender and readers, etc. That is not a threat unique to this software, but affects all internet services these days and is why there was a push for HTTPS, improved TLS protocol versions and end-to-end-encryption over the last decade.

What we can do, for this software, is to minimize the amount of data stored on the server, to reduce these risks. For example, we don't store the paste creation date, only the expiration one. Based on expiration alone, in combination with the list of possible life-times offered by the instance one can only guesstimate the creation time. But in access logs or sniffed traffic the creation times are still discernible.

**Reply by rockman413** [reply]
Thanks a lot.
   So in short, am I correct to think this pivatebin paste which doesn't require a password to open, the content within is still 99.9% safe? 
  And the same as an iCloud link or Office 365 link?

**Reply by elrido** [reply]
I would sure hope so, but as a co-developer of this solution I'm also obviously biased.

**Reply by rockman413** [reply]
I would sure hope so
---> Do you mean you would assume it's safe or it is the case (responding to my question of it should be 99.9% safe)

**Reply by elrido** [reply]
I can not provide you with a formal proof, amongst other things because I do not know the details of the inner workings of the other, proprietary services that you compare it to. I am also obviously biased towards this project, that I collaborate on.

I am under the assumption that our software is secure when used responsibly. The biggest risk, AFAIK, is not that anyone would guess your paste ID and learn about it that way, but rather that your link URL leaks, because it is transmitted via a method that turns out not to be secure enough or made public (on purpose or accidentally). I assume this to also apply to access-links for iCloud and Co. since they use similar mechanisms.

If anyone does find scenarios where this is not the case or has any suggestions on how to improve our softwares' design, I would hope these get [disclosed to us](https://github.com/PrivateBin/PrivateBin/blob/master/SECURITY.md) or reported as an issue, so we can look into it and improve on it.

**Reply by rockman413** [reply]
Ah, understood. Thank you. Just to let you know I'm not asking for a formal proof~ I just want to understand how things work. That's why I said 99.9%,  not 100% safe.

So in order to learn about this and similar cases( like iCloud or office 365 shared link), I understand as: multiple guess or try at different URL will likely trip DDns protection (or take years to try out successfully) so the PrivateBin link should be (like 99%, but of course not 100%) safe.  Am I correct?

**Reply by rockman413** [reply]
Ah, understood. Thank you. Just to let you know I'm not asking for a formal proof~ I just want to understand how things work. That's why I said 99.9%,  not 100% safe.

So in order to learn about this and similar cases( like iCloud or office 365 shared link), I understand as: multiple guess or try at different URL will likely trip DDns protection (or take years to try out successfully) so the PrivateBin link should be (like 99%, but of course not 100%) safe.  Am I correct?

**Reply by rugk** [reply]
See also https://github.com/PrivateBin/PrivateBin/wiki/FAQ#if-i-use-privatebin-and-post-the-link-in-a-forum-in-social-media-etc-is-my-content-secure and https://github.com/PrivateBin/PrivateBin/wiki/FAQ#the-url-is-so-long-cant-i-just-use-an-url-shortener for why URL shortener are discouraged.

The TL;DR is: as long as the url you share is private, the content can be considered secure. If e.g. it is shared publicly in a forum or so, (and it is scraped or manually accessed by a human) obviously no magic can make it "secure".

Also you always have a way to set a password. But similarly to the url: if that is shared publicly or to all who know/have access to it, it is accessible by all of them.
After all, this is the aim of this project. Being a storage for information that can be shared/read by someone. Somehow someone needs to get access. 

The URL can practically not be guessed, as explained before.

**Comment by rugk** [comment]
For a full more detailed security analysis, you can also have a look at our documented threat model: https://github.com/PrivateBin/PrivateBin/wiki/Threat-Model
